### DISCUSSION TITLE
Latency in Redis operations with Elasticache in cross-region.

### DISCUSSION THREAD

**Discussion by awesomejavacoder** [redis/lettuce#2361] [Q&A]
Hello  Lettuce Team👋 

We are facing latency issue while doing all Redis operations. We are using Spring boot with embedded Netty and using Reactive Lettuce API. For Redis, we are using AWS Elasticache, with redis engine 6.x. Here are some observations:

- Reading keys by pattern taking around 7 secs. This pattern resolves to some 50 keys. The code we have is 
```
@Autowired
    private ReactiveRedisOperations<String, String> reactiveRedisOps;

    public Mono<List<String>> readObjects(String pattern){
        return reactiveRedisOps.scan(ScanOptions.scanOptions().match(pattern).build())
                .collectList()
                .flatMap(keys -> reactiveRedisOps.opsForValue().multiGet(keys));
    }
```
 - Making direct calls to the Redis host from within the container is taking less than a second. The commands we have used are: 
```
exec 3<>/dev/tcp/<redis_host_dns_name>/6379 && echo -e "KEYS pattern\r\n" >&3 && head -c 7 <&3
exec 3<>/dev/tcp/<redis_host_dns_name>/6379 && echo -e "MGET K1 K2 ...K50\r\n" >&3 && head -c 7 <&3
```

- The container is able to access other upstream endpoints via WebClient pretty fast. So it doesn't like Netty is resolving slowly.

- I tried to set the logging at TRACE level, but I hardly see any logs from Lettuce library, on how it is establishing connection to the Redis host and reading data etc.
- This only happens when we have container in region us-west-2 and Redis host in region us-east-1. When the container and the Redis host are in the same region, no issues absolutely, and operations are getting done in under a sec.

Appreciate any help and suggestions at your earliest!

**Comment by mp911de** [comment]
You can enable command latency metrics so that you can introspect which command took how long.

**Reply by awesomejavacoder** [reply]
Hi @mp911de I enabled command latency metrics using micrometer and plug those metrics to Grafana. Following was the observations for metrics `lettuce.command.completion`:

Cross-region:
<img width="1716" alt="Redis-Lettuuce Cross-region latency metrics" src="https://user-images.githubusercontent.com/45470789/228738104-73675be7-580c-458b-a8fb-d0accca65bc5.png">

Same-region:
<img width="1714" alt="Redis-Lettuuce Same-region latency metrics" src="https://user-images.githubusercontent.com/45470789/228738137-07af366d-492c-4683-83b8-d623f26b03fa.png">


The commands that I was using mostly was SCAN followed by MGET for these measurements. In cross region, both of these operations together took 7 plus secs and in the same region setup, they took around 200 ms. However, metrics reported far less values for these latencies but still the metrics for cross-region has values which are 7x for SCAN and 100X for MGET when compared with same region values. 
Where I can look further for why cross-region has such a high latency? Appreciate any suggestions!
